client/extensions/woocommerce/state/sites/promotions/handlers.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import { fetchCoupons } from 'woocommerce/state/sites/coupons/actions';
+import { fetchProducts } from 'woocommerce/state/sites/products/actions';
+import {
+	WOOCOMMERCE_PROMOTIONS_REQUEST,
+	WOOCOMMERCE_PRODUCTS_REQUEST_SUCCESS,
+	WOOCOMMERCE_COUPONS_UPDATED,
+} from 'woocommerce/state/action-types';
+
+const debug = debugFactory( 'woocommerce:promotions' );
+
+// How many products to load per page at a time.
+// TODO: Remove this once we have a Promotions API.
+const itemsPerPage = 30;
+
+export default {
+	[ WOOCOMMERCE_PROMOTIONS_REQUEST ]: [ promotionsRequest ],
+	[ WOOCOMMERCE_PRODUCTS_REQUEST_SUCCESS ]: [ productsRequestSuccess ],
+	[ WOOCOMMERCE_COUPONS_UPDATED ]: [ couponsUpdated ],
+};
+
+export function promotionsRequest( { dispatch }, action ) {
+	const { siteId } = action;
+	const perPage = action.perPage || itemsPerPage;
+
+	// To get the list of promotions, we have to merge the list of
+	// products that are "on sale" and all coupons, and sort them
+	// by expiration date. The WooCommerce API doesn't support ordering
+	// by expiration date, so for now, we have to get the entire lists
+	// and sort them ourselves.
+	//
+	// In the future, we will either modify the API to support sorting this way,
+	// or even better, create a Promotions API to handle this on the server.
+	//
+	// TODO: Update after API support changes, one way or the other.
+
+	// Have to fetch all products, because we need not just the ones on sale,
+	// but the ones who have sale dates which are expired or still to come.
+	dispatch( fetchProducts( siteId, { offset: 0, per_page: perPage } ) );
+
+	// Fetch all coupons as well.
+	dispatch( fetchCoupons( siteId, { offset: 0, per_page: perPage } ) );
+}
+
+export function productsRequestSuccess( { dispatch }, action ) {
+	const { siteId, products, params, totalProducts } = action;
+
+	if ( undefined !== params.offset ) {
+		debug( `Products ${ params.offset + 1 }-${ params.offset + products.length } out of ${ totalProducts } received.` );
+
+		const remainder = totalProducts - params.offset - products.length;
+		if ( remainder ) {
```

Comment:
So whenever the promotion page loads, we are going to request keep making requests against the products + coupons endpoints until all pages have been fetched?

Reply:
That's correct. It's really the only way to be able to combine the two lists of items.

Reply:
😨 that could lead to quite a bit of network chatter on sites with a large quantity of products. Is there any other query params we can add ( too bad about the sale flag ) to limit this dataset further?

Reply:
I've looked at the options. Our API blocks any meta filtering, so we can't really filter any other way. This really needs a Promotions API to do this right. Perhaps we should consider adding that API to wc-api-dev sooner rather than later. In the meantime, we can tweak the perPage setting to strike a balance between number of requests and data per request.

Either way, I'd like for this to be somewhat temporary until we get a proper Promotions API in place.